protocols/identify/src/protocol.rs

```diff
@@ -193,7 +193,7 @@ async fn recv<T>(mut socket: T) -> Result<Info, UpgradeError>
 where
     T: AsyncRead + AsyncWrite + Unpin,
 {
-    socket.close().await?;
+    socket.flush().await?;
```

Comment:
Is flushing needed here? Would removing the whole line be enough?

See https://github.com/libp2p/rust-libp2p/issues/3298#issuecomment-1369055550 for more details.

Reply:
Does it work if we ignore the error on the close? Would that be a viable hot-fix too?

Plus, I'd like us to have a comment here linking to an issue on GitHub tracking the overall problem.

Reply:
> Does it work if we ignore the error on the close? Would that be a viable hot-fix too?

Preference for removing the close as it is not needed in the first place. Ignoring an error might hinder debugging in the future.

> Plus, I'd like us to have a comment here linking to an issue on GitHub tracking the overall problem.

Yes please. Good idea.

Reply:
@elenaf9 friendly ping.

Reply:
After looking into this a bit more I think closing a stream here is / will remain problematic, independently of #3343. 
The reason is that we also use `recv` in the inbound upgrade on `Push`. If this is combined with using upgrade `Version::V1Lazy` I think the following can happen in theory:
1. The dialer opens a new outbound identify-push stream
2. With the `V1Lazy` it directly flushes the identify payload together with the protocol-id
3. It drops the stream once the write is done
4. The listener reads the protocol-id and echoes it back before starting the inbound upgrade
5. the remote never reads the echoed-back protocol-id, thus closing the stream fails

In go-libp2p this issue is solved by simply ignoring the error in both, the multi-stream-select upgrade[^1] and when reading the identify[^2] response.

I am not too familiar with multi-stream select in rust-libp2p so it could be that in the above case the dialer would not drop the stream until it received the echoed-back protocol-id. That said, even if it's implemented like that in rust-libp2p, I assume from the linked go-libp2p code that this is how they are doing it. So I think we also have to handle that case. // CC @MarcoPolo since we talked about this out-of-band - did I understand it correctly?

Am I missing something?

[^1]: https://github.com/multiformats/go-multistream/blob/5555a5c43be95c669ce5637723ccf690a6a30e23/multistream.go#L203-L205
[^2]: https://github.com/libp2p/go-libp2p/blob/6b9c11680e0bb412ef7515f266531c23245015dc/p2p/protocol/identify/id.go#L435

Reply:
> the remote never reads the echoed-back protocol-id, thus closing the stream fails

I am not sure I understand the last step here. Note that closing a stream is a local operation only in TCP + Yamux. It only flushes the stream buffers to the local connection buffers. It does not ensure that the closing of the stream is communicated to the remote.

https://github.com/libp2p/rust-yamux/blob/72ccd5734f71971fc1f02ac4a5f43b8eb4dff660/yamux/src/connection/stream.rs#L365-L387

> I am not too familiar with multi-stream select in rust-libp2p so it could be that in the above case the dialer would not drop the stream until it received the echoed-back protocol-id. 

It should await receiving the protocol-id-echo.

https://github.com/libp2p/rust-libp2p/blob/ab59af4d46e0f6835429204f622475adcb6b74df/misc/multistream-select/src/negotiated.rs#L308-L310

Never the less, I am advocating for removing the `close` in `recv`. Any objections @elenaf9?

Reply:
>>  the remote never reads the echoed-back protocol-id, thus closing the stream fails
>
> I am not sure I understand the last step here. Note that closing a stream is a local operation only in TCP + Yamux. It only flushes the stream buffers to the local connection buffers. It does not ensure that the closing of the stream is communicated to the remote. Communicating the closing to the remote is asynchronous to calling close. The connection task will make sure the closing of the stream is communicated to the remote.

Sorry, I was probably not clear with what I meant; I was referring to @thomaseizinger suggestion to add a comment here that links to #3343. I was arguing that **because** in QUIC (contrary to TCP+Yamux) the close is not local only but instead fails if not all data gets acknowledged, even with #3343 resolved the close call in QUIC could cause issues in the described scenario. 

I was mostly wondering if my understanding was correct. If it is, I would add a comment that describes that. 

> Never the less, I am advocating for removing the `close` in `recv`. Any objections @elenaf9?

No objections; I will remove it.

Reply:
> Edit: What do you think of https://github.com/libp2p/rust-libp2p/commit/cf21edba36c8faa87b6de09284fbc596d33c162f? 

Friendly ping @mxinden @thomaseizinger.

Reply:
I am okay with the comment. At the end of the day, this is a temporary situation until we fix `quinn` (+ `quinn-proto`), right?

Reply:
Not sure. That's what I meant above: even with the fix, it could cause errors in case of identify `Push` + `V1Lazy` if the remote doesn't read the echo-ed back protocol-id. Not an issue if both peers are rust-libp2p, but I am not sure how go-libp2p implements it.